py/dml/ctree.py

```diff
@@ -4297,6 +4307,191 @@ def read(self):
+
+class TemplatesSubRef(NonValue):
+    '''A subreference of the 'templates' param of an object, specifying the
+       template of the template-qualified method implementation call'''
+    @auto_init
+    def __init__(self, site, templates_ref, template_name): pass
+
+    def __str__(self):
+        return f'{self.templates_ref}.{self.template_name}'
+
+def mkTemplatesSubRef(site, templates_ref, template_name):
+    tmpl = dml.globals.templates.get(template_name)
+    if tmpl is None:
+        raise ENTMPL(site, template_name)
+    if tmpl not in templates_ref.node.templates:
+        raise ETQMIC(site, templates_ref.node.ident, template_name)
+    return TemplatesSubRef(site, templates_ref, template_name)
+
+def mkTemplateQualifiedMethodRef(site, templates_subref, method_name):
+    node = templates_subref.templates_ref.node
+    indices = templates_subref.templates_ref.indices
+
+    template_name = templates_subref.template_name
+
+    rank_to_candidate = {}
+
+    # Add any shared method implementation as a possible candidate
+    trait = dml.globals.traits.get(template_name)
+    shared_method = None
+    if trait is not None:
+        for impl_trait in trait.method_impl_traits.get(method_name, []):
+            impl_template = dml.globals.templates[impl_trait.name]
+            shared_method = impl_trait.method_impls[method_name]
+            rank_to_candidate[impl_template.spec.rank] = (
+                impl_template, shared_method)
+
+
+    studied_templates = set()
+    pending_templates = Set((dml.globals.templates[template_name],)
+                            # don't bother going through regular methods
+                            # if if some shared method impl. is
+                            # unoverridable (can only be one if so)
+                            * (shared_method is None
+                               or shared_method.overridable))
+
+    # Find the highest ranking candidate of each distinct template hierarchy
+    # that is actually implemented for the object
+    while pending_templates:
+        tmpl = pending_templates.pop()
+        studied_templates.add(tmpl)
+        (final, implements_method,
+         next_templates) = tmpl.get_potential_method_impl_details(method_name)
+
+        if final:
+            rank_to_candidate.clear()
+            if next_templates:
+                [tmpl] = next_templates
+            elif not implements_method:
+                break
+
+            candidate = node.template_method_impls.get((tmpl, method_name))
+            assert candidate is not None
+            rank_to_candidate[tmpl.spec.rank] = (tmpl, candidate)
+            break
+
+        if implements_method:
+            candidate = node.template_method_impls.get((tmpl, method_name))
+            if candidate is not None:
+                rank_to_candidate[tmpl.spec.rank] = (tmpl, candidate)
+                continue
+
+        pending_templates.update(filter(lambda c: c not in studied_templates,
```

Comment:
`update(c for c in next_remplates if c not in studied_templates)` ?

Reply:
I suppose it is ever-so-slightly less verbose... goddamit `filter` is so hard to find an excuse to use

Reply:
Irrelevant now with other changes.